asyncapi/shopify/dispatcher_service.bal

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/crypto;
+import ballerina/http;
+import ballerina/log;
+import ballerinax/asyncapi.native.handler;
+
+service class DispatcherService {
+    *http:Service;
+    private map<GenericServiceType> services = {};
+    private handler:NativeHandler nativeHandler = new ();
+    private ListenerConfig listenerConfig;
+
+    public function init(ListenerConfig listenerConfig) {
+        self.listenerConfig = listenerConfig;
+    }
+
+    isolated function addServiceRef(string serviceType, GenericServiceType genericService) returns error? {
+        if self.services.hasKey(serviceType) {
+            return error("Service of type " + serviceType + " has already been attached");
+        }
+        self.services[serviceType] = genericService;
+    }
+
+    isolated function removeServiceRef(string serviceType) returns error? {
+        if !self.services.hasKey(serviceType) {
+            return error("Cannot detach the service of type " + serviceType + 
+                ". Service has not been attached to the listener before");
+        }
+        _ = self.services.remove(serviceType);
+    }
+
+    // We are not using the (@http:payload GenericEventWrapperEvent g) notation because of a bug in Ballerina.
+    // Issue: https://github.com/ballerina-platform/ballerina-lang/issues/32859
+    resource function post .(http:Caller caller, http:Request request) returns error? {
+        byte[] binaryPayload = check request.getBinaryPayload();
+        json payload = check request.getJsonPayload();
+        string eventName = check request.getHeader("x-shopify-topic");
+        if request.hasHeader("X-Shopify-Hmac-SHA256") {
+            string hmacHeader = check request.getHeader("X-Shopify-Hmac-SHA256");
+            byte[] digest = check crypto:hmacSha256(binaryPayload, self.listenerConfig.apiSecretKey.toBytes());
+            string computedHmac = digest.toBase64();
+            if computedHmac != hmacHeader {
+                // Validate API secret key with X-Shopify-Hmac-SHA256 header for intent verification
+                log:printError("Signature verification failure!");
+                return caller->respond(http:STATUS_UNAUTHORIZED);
```

Comment:
Does shopify expects 401 status code as the response?. If so we need to add a status code also. ex :- https://github.com/ballerina-platform/asyncapi-triggers/blob/main/asyncapi/github/dispatcher_service.bal#L61

Reply:
This is resolved now and amended the changes.